compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3053,6 +3053,11 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                 return;
             }
 
+            // Scalars which have invalid values cannot be undef.
+            if !scalar.is_always_valid(self) {
+                attrs.set(ArgAttribute::NoUndef);
```

Comment:
With this logic, can't we remove some of the logic from https://github.com/rust-lang/rust/pull/93670 since those types also are "not always valid"?

Reply:
I initially thought the same, but then we don't put `noundef` on `Option<&T/&mut T/Box<T>>>`, since they have no invalid values.

Reply:
Oh, I didn't realize those would get the attribute.

OTOH that still means we are missing `Option<NonZero*>`, so maybe there's a more general way to handle this by special-casing enums? Like, if the scalar has `Multiple` layout, then that one scalar contains its discriminant and hence we can definitely set `NoUndef`? (This might miss newtypes around such enums, not sure if that's a problem.)

Reply:
You're right.

I'm a bit hesistant to add more complex logic here. I think the missing cases are fairly rare, just `Option<NonZero*/NonNull>` (and equivalent types), and enums with exactly 256 (or 65536, etc.) variants. In the long run we can switch to the `undef` tracking added by #94527 and handle these cases without having to inspect the shape of the layout at all.

Reply:
Sure. :) I just hope we won't forget to clean up that duplication of logic in two different places ;)